buildSrc/src/main/kotlin/io/spine/internal/gradle/publish/SpinePublishing.kt

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright 2022, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.internal.gradle.publish
+
+import io.spine.internal.gradle.Repository
+import org.gradle.api.Project
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.findByType
+
+/**
+ * Configures [SpinePublishing] extension.
+ *
+ * This extension sets up publishing of artifacts to Maven repositories.
+ *
+ * The extension can be configured for single- and multi-module projects.
+ *
+ * When used with a multi-module project, the extension should be opened in a root project's
+ * build file. The published modules are specified explicitly by their names:
+ *
+ * ```
+ * spinePublishing {
+ *     modules = setOf(
+ *         "subprojectA",
+ *         "subprojectB",
+ *     )
+ *     destinations = targetRepositories {
```

Comment:
As I already commented separately, can we have it simply `targetRepositories = setOf(...)`?

The problem with the current code is that we say the same thing twice, but with different words.

Reply:
`targetRepositories`, previously known as `spineRepositories` is just a complementary method. The method exposes ` PublishingRepos` object, which is set of pre-defined repositories. Two code snippets below are identical.

 ```
destinations = setOf(
    PublishingRepos.cloudRepo.destination,
    PublishingRepos.cloudArtifactRegistry,
    PublishingRepos.gitHub("base"),
)
```

```
destinations = targetRepositories {
        setOf(
            cloudRepo.destination,
            cloudArtifactRegistry,
            gitHub("base"),
        )
    }
```

Now, the extension is allowed to accept any `Set<Repository>`, not only the ones from `PublishingRepos`.

 I see several options to eliminate the ambiguity:

1. Drop the complementary method. We can live with several `PublishingRepos` in a column.
2. Leave it as `spineRepositories`. Think of it just as a shortcut for `PublishingRepos` object.
3. Drop `destinations` property. Create a function `fun destinations(select: PublishingRepos.() -> Set<Repository>)`. This would make the extension to accept repositories only from `PublishingRepos`.
4. Similar to (3), but with overloading:

```
fun destinations(select: () -> Set<Repository>)
fun destinations(select: PublishingRepos.() -> Set<Repository>)
```